### ISSUE TITLE
0.11 fails to build on Linux i686 (32-bit)

### ISSUE THREAD

**Issue by felixonmars** [haskell-crypto/cryptonite#67]
I didn't get the error with 0.10, so this seems a new problem that first appears in 0.11:

```
cbits/ed448/x448.c:21:9:
     error: unknown type name '__uint128_t'
     typedef __uint128_t decaf_dword_t;
             ^

cbits/ed448/x448.c:22:9:
     error: unknown type name '__int128_t'
     typedef __int128_t decaf_sdword_t;
             ^
cbits/ed448/x448.c: In function 'gf_mul':

cbits/ed448/x448.c:83:48:
     warning: right shift count >= width of type [-Wshift-count-overflow]
         accum[X448_LIMBS-1] += accum[X448_LIMBS-2] >> LBITS;
                                                    ^

cbits/ed448/x448.c:85:48:
     warning: right shift count >= width of type [-Wshift-count-overflow]
         accum[X448_LIMBS/2] += accum[X448_LIMBS-1] >> LBITS;
                                                    ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:8:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
            ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:16:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                    ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:24:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                            ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:32:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                                    ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:40:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                                            ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:48:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                                                    ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:56:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                                                            ^

cbits/ed448/x448.c:87:45:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             accum[j] += accum[(j-1)%X448_LIMBS] >> LBITS;
                                                 ^

cbits/ed448/x448.c:50:64:
     note: in definition of macro 'FOR_LIMB_U'
            op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; op;i++; \
                                                                    ^
cbits/ed448/x448.c: In function 'gf_canon':

cbits/ed448/x448.c:186:15:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             carry >>= LBITS;
                   ^

cbits/ed448/x448.c:64:76:
     note: in definition of macro 'FOR_LIMB'
     #define FOR_LIMB(i,op) { unsigned int i=0; for (i=0; i<X448_LIMBS; i++)  { op; }}
                                                                                ^

cbits/ed448/x448.c:196:15:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             carry >>= LBITS;
                   ^

cbits/ed448/x448.c:64:76:
     note: in definition of macro 'FOR_LIMB'
     #define FOR_LIMB(i,op) { unsigned int i=0; for (i=0; i<X448_LIMBS; i++)  { op; }}
                                                                                ^
cbits/ed448/x448.c: In function 'gf_deser':

cbits/ed448/x448.c:207:91:
     warning: right shift count >= width of type [-Wshift-count-overflow]
             for (bits += 8; (bits>=LBITS || i==X448_BYTES-1) && k<X448_LIMBS; bits-=LBITS, buf>>=LBITS) {
                                                                                               ^

cbits/ed448/x448.c:213:59:
     warning: right shift count >= width of type [-Wshift-count-overflow]
         FOR_LIMB(i, accum = (accum + s->limb[i] - P->limb[i]) >> WBITS );
                                                               ^

cbits/ed448/x448.c:64:76:
     note: in definition of macro 'FOR_LIMB'
     #define FOR_LIMB(i,op) { unsigned int i=0; for (i=0; i<X448_LIMBS; i++)  { op; }}
                                                                                ^
```

**Comment by tekul** [comment]
Looks like it's caused by the new C code from #66. Perhaps @centromere can add a fix?

**Comment by vincenthz** [comment]
probably "fixed" by ec130aeca009ec94cc1f3c119a282bb57f6577f2; probably should get the bitness out of a better cpp macro than ARCH_X86_64

@felixonmars could you check that the master repository is building now ?

**Comment by felixonmars** [comment]
Works like a charm here. Thanks :)

**Comment by gaa-cifasis** [comment]
Also confirmed the fix. +1 for a hackage release

**Comment by danielwilson654** [comment]
Thanks for posting the fix! I have been trying to compile pandoc 1.16 to a 32bit executable. How do I import the fix into stack? I ran "stack setup" in a new user to try and download the fix but the installation still fails.

I may try and use the Haskell platform instead. Can a run a "cable import" to download the fix? I tried to use cable a few days ago but encountered the error mentioned above.

I am new to compiling as you may have guessed! Any help will be greatly appreciated!

**Comment by danielwilson654** [comment]
Sorry, the above command should read "cabal update". I attempted this method but the result was still the same. do I need to download the fix from git repo?

terminal output:
Failed to install cryptonite-0.11
cabal: Error: some packages failed to install:
connection-0.2.5 depends on cryptonite-0.11 which failed to install.
cryptonite-0.11 failed during the building phase. The exception was:
ExitFailure 1

**Comment by ubuntu-user1** [comment]
Thanks for posting the fix! I have been trying to compile pandoc 1.16 to a 32bit executable. How do I import the fix into stack? I ran "stack setup" in a new user to try and download the fix but the installation still fails.

I tried to use the Haskell platform instead. I ran "cabal install" to download the fix but I got the same error when compiling as I did a few days ago. Do I need to download the fix from git repo?

terminal output:
Failed to install cryptonite-0.11
cabal: Error: some packages failed to install:
connection-0.2.5 depends on cryptonite-0.11 which failed to install.
cryptonite-0.11 failed during the building phase. The exception was:
ExitFailure 1

I am new to compiling as you may have guessed! Any help will be greatly appreciated!

**Comment by tekul** [comment]
@ubuntu-user1 There hasn't been a release with the changes yet, so yes you would need to install from source until the next release is out. It also looks like pandoc has a build flag to disable HTTPS, which you could use. I'd guess this is where the dependency comes from.

**Comment by vincenthz** [comment]
Thanks everyone reporting success. I've released 0.12 to hackage now
